ddtrace/tracer/sampler.go

```diff
@@ -147,3 +154,273 @@ func (ps *prioritySampler) apply(spn *span) {
+	} else {
+		span.SetTag(ext.SamplingPriority, ext.PriorityAutoReject)
+	}
+	span.SetTag(keyRulesSamplerLimiterRate, rate)
+}
+
+// SamplingRule is used for applying sampling rates to spans that match
+// the service name, operation name or both.
+// For basic usage, consider using the helper functions ServiceRule, NameRule, etc.
+type SamplingRule struct {
+	Service *regexp.Regexp
+	Name    *regexp.Regexp
+	Rate    float64
+
+	exactService string
+	exactName    string
+}
+
+// ServiceRule returns a SamplingRule that applies the provided sampling rate
+// to spans that match the service name provided.
+func ServiceRule(service string, rate float64) SamplingRule {
+	return SamplingRule{
+		exactService: service,
+		Rate:         rate,
+	}
+}
+
+// NameRule returns a SamplingRule that applies the provided sampling rate
+// to spans that match the operation name provided.
+func NameRule(name string, rate float64) SamplingRule {
+	return SamplingRule{
+		exactName: name,
+		Rate:      rate,
+	}
+}
+
+// NameServiceRule returns a SamplingRule that applies the provided sampling rate
+// to spans matching both the operation and service names provided.
+func NameServiceRule(name string, service string, rate float64) SamplingRule {
+	return SamplingRule{
+		exactService: service,
+		exactName:    name,
+		Rate:         rate,
+	}
+}
+
+// RateRule returns a SamplingRule that applies the provided sampling rate to all spans.
+func RateRule(rate float64) SamplingRule {
+	return SamplingRule{
+		Rate: rate,
+	}
+}
+
+// match returns true when the span's details match all the expected values in the rule.
+func (sr *SamplingRule) match(s *span) bool {
+	if sr.Service != nil && !sr.Service.MatchString(s.Service) {
+		return false
+	} else if sr.exactService != "" && sr.exactService != s.Service {
+		return false
+	}
+	if sr.Name != nil && !sr.Name.MatchString(s.Name) {
+		return false
+	} else if sr.exactName != "" && sr.exactName != s.Name {
+		return false
+	}
+	return true
+}
+
+// samplingLimiter is used to limit the quantity of spans sampled by the rules sampler
+// and internally maintains measurements for "effective rate".
+type samplingLimiter struct {
+	limiter *rate.Limiter
+
+	// "effective rate" calculations
+	mu            sync.Mutex // guards below fields
+	currentPeriod time.Time  // timestamp, to detect when counters need resetting
+	previousRate  float64    // previous second's rate, averaged with current rate for smoothing
+	allowed       int        // number of spans allowed by rate limiter in the current period
+	seen          int        // number of spans checked by rate limiter in the current period
+}
+
+// samplingDecision returns the rate limiter's decision to allow the span to be sampled, and the
+// effective rate at the time it is called. The effective rate is computed by averaging the rate
+// for the previous second with the current rate
+func (r *samplingLimiter) samplingDecision(now time.Time) (bool, float64) {
```

Comment:
Why not "allow" or "allowOne"? That's exactly what it is and what it does?
```suggestion
func (r *samplingLimiter) allow(now time.Time) (bool, float64) {
```